lib/ecto/query/planner.ex

```diff
@@ -814,6 +849,30 @@ defmodule Ecto.Query.Planner do
     {Enum.reverse(exprs), counter}
   end
 
+  defp validate_and_increment(:with, _query, nil, counter, _operation, _adapter) do
+    {nil, counter}
+  end
+
+  defp validate_and_increment(:with, _query, with_expr, counter, operation, adapter) do
+    fun = &validate_and_increment(&1, &2, &3, &4, operation, adapter)
+
+    {queries, counter} =
+      Enum.reduce with_expr.queries, {[], counter}, fn
+        {name, %Ecto.Query{} = query}, {queries, counter} ->
+          {query, counter} = traverse_exprs(query, operation, counter, fun)
+          {query, _} = normalize_select(query)
+          {_, select} = query |> subquery_select(adapter)
+          keys = %{select: select} |> subquery_types() |> Keyword.keys()
+          query = query.select.fields |> update_in(&Enum.zip(keys, &1))
+          {[{name, query} | queries], counter}
+
+        other, {queries, counter} ->
+          {[other | queries], counter}
```

Comment:
I think we need to call `fun` recursively here too so we handle any argument in the fragment.